File: tuneable-mobile/modules/now-playing/src/NowPlayingModule.ts
import { NativeModule, requireNativeModule } from 'expo-modules-core';

export type NowPlayingCommand =
  | 'play'
  | 'pause'
  | 'toggle'
  | 'next'
  | 'previous'
  | 'seek'
  | 'skipForward'
  | 'skipBack';

export type NowPlayingCommandEvent = {
  command: NowPlayingCommand;
  positionMs?: number;
  intervalMs?: number;
};

export type NowPlayingParams = {
  title: string;
  artist: string;
  albumTitle?: string;
  artworkUrl?: string | null;
  duration: number;
  elapsed: number;
  playbackRate: number;
  isPlaying: boolean;
  mode: 'music' | 'podcast';
};

declare class NowPlayingNativeModule extends NativeModule<{
  onCommand: (event: NowPlayingCommandEvent) => void;
}> {
  setNowPlaying(params: NowPlayingParams): void;
  updateElapsed(elapsed: number, duration: number, playbackRate: number): void;
  clear(): void;
}

export default requireNativeModule<NowPlayingNativeModule>('NowPlaying');

File: tuneable-mobile/modules/now-playing/android/src/main/java/expo/modules/nowplaying/NowPlayingModule.kt
package expo.modules.nowplaying

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class NowPlayingParams : Record {
  @Field var title: String = "Tuneable"
  @Field var artist: String = ""
  @Field var albumTitle: String = "Tuneable"
  @Field var artworkUrl: String? = null
  @Field var duration: Double = 0.0
  @Field var elapsed: Double = 0.0
  @Field var playbackRate: Double = 1.0
  @Field var isPlaying: Boolean = false
  @Field var mode: String = "music"
}

class NowPlayingModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NowPlaying")
    Events("onCommand")

    // Lock-screen Now Playing is implemented natively on iOS. Android keeps
    // these as no-ops so adding the module cannot affect app startup.
    Function("setNowPlaying") { _: NowPlayingParams -> }
    Function("updateElapsed") { _: Double, _: Double, _: Double -> }
    Function("clear") { }
  }
}
